src/dataflow/src/server.rs

```diff
@@ -185,31 +185,41 @@ pub enum Command {
 }
 
 impl Command {
-    /// Produces a copy of the command suitable for the indicated worker.
+    /// Partitions the command among `peers` many workers.
     ///
     /// This is used to subdivide commands that can be sharded across workers,
     /// for example the `plan::Constant` stages of dataflow plans, and the
     /// `Command::Insert` commands that may contain multiple updates.
-    pub fn clone_for_worker(&self, index: usize, peers: usize) -> Self {
-        match self {
-            Command::CreateDataflows(dataflows) => {
-                Command::CreateDataflows(
-                    dataflows
-                        .iter()
-                        .map(|dataflow| {
-                            // We do this here, because it is hard to have `dataflow_types::Dataflow` know about
-                            // `dataflow::Plan`.
-                            // Each dataflow we construct should shard its `Constant` collections.
-                            let objects_to_build = dataflow
-                                .objects_to_build
-                                .iter()
-                                .map(|description| dataflow_types::BuildDesc {
-                                    id: description.id,
-                                    view: description.view.clone_for_worker(index, peers),
-                                })
-                                .collect::<Vec<_>>();
-                            // Clone all fields, other than `objects_to_build` defined above.
-                            DataflowDescription {
+    pub fn partition_among(self, parts: usize) -> Vec<Self> {
```

Comment:
I'm not clear about the cost of returning `Vec<Self>`. It seems that in most cases, we return a vector, which we immediately turn into an iterator and drop. Now this might not be much of a concern since it's small vectors mostly, but still it's some work.
Do you think it's worth going over this change and providing a `&mut Vec<Self>` to the function to write the plan into, or return a `Box<dyn Iterator<Item=Plan>>`?

Reply:
The comments applies to `render/mod.rs` just as much, maybe even more.

Reply:
I think it wouldn't be transformative to do this. Each of the things *in* the `Vec` need to be put in their own `Box` just after receiving the `Vec`, and so while it would be cheaper to avoid the allocation with iterators, it would be taking `workers + 1` allocations down to `workers`. If it tidies the code at all it could make sense; I'll look for that.

Reply:
And, I think each `Box<dyn Iterator<Item=Plan>>` would be an allocation itself, so maybe net effect would be swapping out one `Vec` sized allocation for a fixed sized allocation?

Reply:
That seems correct. In case a switch arm would create a new iterator by means of ingesting another one we would push the whole chain of iterators to the heap, which would be of dubious benefit. So let's go with the current approach!